File: src/main/java/shopping/product/application/ProductService.java
package shopping.product.application;

import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import shopping.product.application.dto.ProductCreateRequest;
import shopping.product.application.dto.ProductUpdateRequest;
import shopping.product.domain.Product;
import shopping.product.domain.ProductRepository;
import shopping.product.domain.ProfanityChecker;
import shopping.product.exception.NotFoundProductException;

@Service
@RequiredArgsConstructor
public class ProductService {

    private final ProfanityChecker profanityChecker;
    private final ProductRepository productRepository;

    public Long create(ProductCreateRequest request) {
        final Product product = new Product(request.name(),
                profanityChecker,
                request.price(),
                request.image());
        final Product saved = productRepository.save(product);
        return saved.getId();
    }

    public void update(Long productId, ProductUpdateRequest request) {
        final Product product = findProduct(productId);
        product.update(request.name(),
                profanityChecker,
                request.price(),
                request.image());
        productRepository.save(product);
    }

    public void delete(Long id) {
        final Product product = findProduct(id);
        productRepository.delete(product);
    }

    public Product findProduct(Long id) {
        return productRepository.findById(id)
                .orElseThrow(() -> new NotFoundProductException("해당 상품을 찾을 수 없습니다. " + id));
    }

    public List<Product> findProducts(List<Long> ids) {
        return productRepository.findByIdIn(ids);
    }
}

File: src/main/kotlin/shopping/product/application/ProductService.kt
package shopping.product.application

import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import shopping.product.domain.BadWordValidator
import shopping.product.domain.ProductNameContainBadWordException
import shopping.product.domain.ProductRepository

@Service
class ProductService(
    private val productRepository: ProductRepository,
    private val badWordValidator: BadWordValidator,
) {
    fun createProduct(request: CreateProductRequest): CreateProductResponse {
        checkProductNameIsNotBadWord(request.name)

        val product = productRepository.save(request.toProduct())

        return CreateProductResponse(product)
    }

    fun updateProduct(
        productId: Long,
        request: UpdateProductRequest,
    ): UpdateProductResponse {
        val product = productRepository.findByIdOrNull(productId) ?: throw ProductNotFoundException(productId)

        checkProductNameIsNotBadWord(request.name)

        product.update(
            name = request.name,
            price = request.price,
            imageUrl = request.imageUrl,
        )
        productRepository.save(product)

        return UpdateProductResponse(product)
    }

    fun getProduct(productId: Long): ProductDetailResponse {
        val product = productRepository.findByIdOrNull(productId) ?: throw ProductNotFoundException(productId)

        return ProductDetailResponse(product)
    }

    fun deleteProduct(productId: Long) {
        val product = productRepository.findByIdOrNull(productId) ?: throw ProductNotFoundException(productId)
        productRepository.delete(product)
    }

    private fun checkProductNameIsNotBadWord(name: String) {
        if (badWordValidator.isBadWord(name)) {
            throw ProductNameContainBadWordException(name)
        }
    }
}
